src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -868,98 +1128,1238 @@ grpc_subchannel_call *grpc_client_channel_get_subchannel_call(
+  }
+  if (pending->batch->send_message) {
+    calld->pending_send_message = false;
+  }
+  if (pending->batch->send_trailing_metadata) {
+    calld->pending_send_trailing_metadata = false;
+  }
+  pending->batch = NULL;
+}
+
 // This is called via the call combiner, so access to calld is synchronized.
 static void fail_pending_batch_in_call_combiner(grpc_exec_ctx *exec_ctx,
                                                 void *arg, grpc_error *error) {
-  call_data *calld = (call_data *)arg;
-  if (calld->waiting_for_pick_batches_count > 0) {
-    --calld->waiting_for_pick_batches_count;
-    grpc_transport_stream_op_batch_finish_with_failure(
-        exec_ctx,
-        calld->waiting_for_pick_batches[calld->waiting_for_pick_batches_count],
-        GRPC_ERROR_REF(error), calld->call_combiner);
-  }
+  grpc_transport_stream_op_batch *batch = (grpc_transport_stream_op_batch *)arg;
+  call_data *calld = (call_data *)batch->handler_private.extra_arg;
+  grpc_transport_stream_op_batch_finish_with_failure(
+      exec_ctx, batch, GRPC_ERROR_REF(error), calld->call_combiner);
 }
 
 // This is called via the call combiner, so access to calld is synchronized.
-static void waiting_for_pick_batches_fail(grpc_exec_ctx *exec_ctx,
-                                          grpc_call_element *elem,
-                                          grpc_error *error) {
+// If yield_call_combiner is true, assumes responsibility for yielding
+// the call combiner.
+static void pending_batches_fail(grpc_exec_ctx *exec_ctx,
+                                 grpc_call_element *elem, grpc_error *error,
+                                 bool yield_call_combiner) {
   call_data *calld = (call_data *)elem->call_data;
   if (GRPC_TRACER_ON(grpc_client_channel_trace)) {
+    size_t num_batches = 0;
+    for (size_t i = 0; i < GPR_ARRAY_SIZE(calld->pending_batches); ++i) {
+      if (calld->pending_batches[i].batch != NULL) ++num_batches;
+    }
     gpr_log(GPR_DEBUG,
             "chand=%p calld=%p: failing %" PRIuPTR " pending batches: %s",
-            elem->channel_data, calld, calld->waiting_for_pick_batches_count,
-            grpc_error_string(error));
+            elem->channel_data, calld, num_batches, grpc_error_string(error));
+  }
+  grpc_transport_stream_op_batch
+      *batches[GPR_ARRAY_SIZE(calld->pending_batches)];
+  size_t num_batches = 0;
+  for (size_t i = 0; i < GPR_ARRAY_SIZE(calld->pending_batches); ++i) {
+    pending_batch *pending = &calld->pending_batches[i];
+    grpc_transport_stream_op_batch *batch = pending->batch;
+    if (batch != NULL) {
+      batches[num_batches++] = batch;
+      pending_batch_clear(calld, pending);
+    }
   }
-  for (size_t i = 0; i < calld->waiting_for_pick_batches_count; ++i) {
-    GRPC_CLOSURE_INIT(&calld->handle_pending_batch_in_call_combiner[i],
-                      fail_pending_batch_in_call_combiner, calld,
+  for (size_t i = yield_call_combiner ? 1 : 0; i < num_batches; ++i) {
```

Comment:
Why do we skip the first batch if we are yielding the call combiner?

I read a bit further on, and will amend the question. Why do we ensure that we fail initial metadata last?